src/software/ai/hl/stp/tactic/tactic.h

```diff
@@ -52,36 +75,60 @@ class Tactic
      *
      * @return true if the Tactic is done and false otherwise
      */
-    virtual bool done() const = 0;
+    bool done() const
+    {
+        bool is_done = false;
+        if (last_execution_robot.has_value())
+        {
+            is_done = fsm_map.at(last_execution_robot.value())->is(boost::sml::X);
+        }
+        return is_done;
+    }
 
     /**
      * Gets the FSM state of the tactic
      *
      * @return the FSM state
      */
-    virtual std::string getFSMState() const = 0;
+    std::string getFSMState() const
+    {
+        std::string state_str = "";
+        if (last_execution_robot.has_value())
+            state_str =
+                getCurrentFullStateName(*fsm_map.at(last_execution_robot.value()));
```

Comment:
Our style guidelines say you should always use braces for clarity, even with single line statements

```suggestion
        if (last_execution_robot.has_value())
        {
            state_str =
                getCurrentFullStateName(*fsm_map.at(last_execution_robot.value()));
        }
```